src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.Unix.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Runtime.InteropServices
+{
+    // Contains UNIX-specific logic for NativeLibrary class.
+    public unsafe sealed partial class NativeLibrary
+    {
+        // [DllImport] (NDirectMethodDesc::FindEntryPoint) doesn't allow lookup by ordinal on non-Windows.
+        private const bool AllowLocatingFunctionsByOrdinal = false;
+
+        // Per PInvokeStaticSigInfo::InitCallConv and GetDefaultCallConv, the default calling convention
+        // on Unix is cdecl.
+        private const CallingConvention FallbackCallingConvention = CallingConvention.Cdecl;
+
+        // In UNIX, the CLR's HINSTANCE is actually a pointer to _MODSTRUCT.
+        // We can extract the underlying OS handle from that structure.
+        private IntPtr OperatingSystemHandle => ((_MODSTRUCT*)_hInstance)->dl_handle;
+
+        private static bool IsValidModuleHandle(IntPtr hModule)
+        {
+            // No validation other than checking for null.
+
+            return (hModule != IntPtr.Zero);
+        }
+
+        // This is a *partial* copy of include/pal/module.h so that we can extract the dl_handle.
+        [StructLayout(LayoutKind.Sequential)]
+        private struct _MODSTRUCT
```

Comment:
I'm not really a fan of this pattern, but this seemed safer than trying to have NativeLibrary.cpp reach deep into the PAL-specific header files. Thoughts?

Reply:
We could expose it as a PAL function, e.g. `PAL_GetDlHandle(HMODULE hMod)` and pinvoke into it here (or expose a wrapper that calls this function as a QCALL) to get rid of this exposure of PAL internals.

It actually seems that moving the IsValidModuleHandle to the native part of the implementation and exposing it as a QCALL would be better.

Reply:
> the IsValidModuleHandle to the native part of the implementation

There is a lot of chattiness between the managed and native parts - there are 5 new FCalls/QCalls already. I think it would be better to have bulk of this implementation in C++, and not have this chattiness.

Reply:
Agree.. I would either all(or mostly) implements in native code or all implements in managed code. but not in middle. It will help interop team maintain these codes in long term..

Reply:
When we were developing this I chatted with quite a few people who have a vested interest in this type's implementation, and the recommendations eventually came down to: (a) reuse as much code as practical; and (b) if new code must be created keep as much of it managed as possible, even at the expense of performance. This would make the code easier to understand and would benefit maintenance in the long run.

In practice the only real implementation logic that exists in C# is a managed equivalent to the C++ `FindEntryPoint` method. Tweaking and reusing the `FindEntryPoint` method directly was dismissed due to how much refactoring it would need. All other C# code is parameter validation / error handling / etc., with p/invokes back into the runtime to do the real work.

If you feel strongly about moving the managed `FindEntryPoint` logic back into C++ let me know and I'll defer to your judgment. Be aware that it will result in larger, more complex, less readable code, with minimal visible effects for consumers of this type.